File: tests/test_emulator_global_attacks.py
"""Native Global WiFi confirmation/active screens backed by offline jobs."""
import unittest
from emulator_browser import browser_session, objects, click_text, click_object

KINDS=[('Blackout','blackout'),('Handshaker','global_handshaker'),('SnifferDog','snifferdog')]

class GlobalAttacksBrowser(unittest.TestCase):
    def setUp(self):
        self.session=browser_session();self.browser,self.url=self.session.__enter__()
        self.addCleanup(self.session.__exit__,None,None,None)
        self.context=self.browser.new_context(viewport={'width':1440,'height':1000})
        self.addCleanup(self.context.close)
        self.page=self.context.new_page();self.errors=[]
        self.page.on('pageerror',lambda e:self.errors.append(str(e)))
        self.page.add_init_script("globalThis.unavailable=[];addEventListener('emulator-unavailable',e=>unavailable.push(e.detail))")

    def ready(self):
        self.page.wait_for_function('globalThis.emulator?.measurements.frames>2')

    def menu(self):
        click_text(self.page,'Global WiFi\nAttacks')

    def load(self,rotation=0):
        self.page.goto(f'{self.url}/?rotation={rotation}');self.ready();self.menu()

    def text(self):return '\n'.join(o['text'] for o in objects(self.page))

    def start(self,label):
        click_text(self.page,label);click_text(self.page,'Yes')
        job=self.page.evaluate('emulator.device.snapshot(emulator.module._emu_current_tab()).active')
        self.assertIsNotNone(job,'Confirmation must start a simulated operation')
        return job

    def stop(self,kind):
        fragment=f'show_{kind}_active_popup.stop_btn'
        click_object(self.page,next(o for o in objects(self.page) if fragment in o['binding']))

    def advance(self):self.page.evaluate('()=>{for(let i=0;i<30;i++)emulator.module._emu_tick(100)}')

    def test_confirm_no_yes_stop_all_operations_and_rotations(self):
        p=self.page
        for rotation in range(4):
            for label,kind in KINDS:
                with self.subTest(rotation=rotation,kind=kind):
                    self.load(rotation)
                    click_text(p,label);click_text(p,'No')
                    self.assertIsNone(p.evaluate('emulator.device.snapshot(0).active'))
                    job=self.start(label);self.advance()
                    self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).operation',job),kind)
                    self.assertIn('Attack in Progress',self.text())
                    if kind=='global_handshaker':
                        result=p.evaluate('(id)=>emulator.device.device.job(id).result.handshakes',job)
                        self.assertGreater(len(result),0)
                        self.assertIn(f'Total: {len(result)}',self.text())
                        self.assertIn(result[-1]['ssid'],self.text())
                    self.stop(kind)
                    self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',job),'completed')
                    self.assertNotIn('Attack in Progress',self.text())
                    self.assertEqual(p.evaluate('unavailable'),[])
        self.assertEqual(self.errors,[])

    def test_disconnect_rejects_start_cancel_reset_and_reopen(self):
        p=self.page
        for label,kind in KINDS:
            with self.subTest(kind=kind):
                self.load()
                p.evaluate("emulator.device.device.setModule('grove',{connected:false})")
                click_text(p,label);click_text(p,'Yes')
                self.assertIsNone(p.evaluate('emulator.device.snapshot(0).active'))
                self.assertIn('disconnected',self.text().lower())
                click_text(p,'No')
                p.evaluate("emulator.device.device.setModule('grove',{connected:true})")
                job=self.start(label)
                p.locator('#cancel').click();self.advance()
                self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',job),'cancelled')
                self.assertNotIn('Attack in Progress',self.text())
                self.stop(kind);self.menu();job=self.start(label)
                p.evaluate("emulator.device.device.setModule('grove',{connected:false})");self.advance()
                self.assertNotIn('Attack in Progress',self.text())
                self.stop(kind)
                with p.expect_navigation():p.locator('#reset').click()
                self.ready();self.menu();self.start(label)
                with p.expect_navigation():p.locator('#reset').click()
                self.ready();self.assertIsNone(p.evaluate('emulator.device.snapshot(0).active'))
                self.assertEqual(p.evaluate('unavailable'),[])
        self.assertEqual(self.errors,[])

    def test_two_modules_have_independent_global_handshake_results_and_stop(self):
        p=self.page;self.load()
        grove=self.start('Handshaker');self.advance()
        click_text(p,'MBUS');self.menu();mbus=self.start('Handshaker')
        self.advance();self.stop('global_handshaker')
        self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',mbus),'completed')
        self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',grove),'running')
        click_text(p,'GROVE');self.advance()
        result=p.evaluate('(id)=>emulator.device.device.job(id).result.handshakes',grove)
        self.assertIn(f'Total: {len(result)}',self.text())
        self.stop('global_handshaker')
        self.assertEqual(p.evaluate('unavailable'),[]);self.assertEqual(self.errors,[])

    def test_busy_refusal_preserves_existing_operation(self):
        p=self.page
        for label,kind in KINDS:
            with self.subTest(kind=kind):
                self.load()
                job=p.evaluate("emulator.device.device.attackStart('grove','karma')")
                click_text(p,label);click_text(p,'Yes')
                self.assertIn('busy',self.text().lower())
                click_text(p,'No')
                self.assertEqual(p.evaluate('emulator.device.snapshot(0).active'),job)
                self.assertEqual(p.evaluate('(id)=>emulator.device.device.job(id).state',job),'running')
                p.evaluate('(id)=>emulator.device.device.cancel(id)',job)
                self.start(label);self.stop(kind)
                self.assertEqual(p.evaluate('unavailable'),[])
        self.assertEqual(self.errors,[])

if __name__=='__main__':unittest.main(verbosity=2)

File: tests/test_emulator_global_attacks.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import {readFile} from 'node:fs/promises';
import {SimulatedDevice} from '../tools/ui_emulator/model/device.mjs';
const seed=JSON.parse(await readFile(new URL('../docs/ui-emulator/neon-district.seed.json',import.meta.url),'utf8'));
const kinds=['blackout','snifferdog','global_handshaker'];

test('global operations need no selection, own one module, advance and freeze on stop',()=>{
 for(const kind of kinds){
  const d=new SimulatedDevice(seed),id=d.attackStart('grove',kind);
  assert.equal(d.job(id).state,'running');
  assert.equal(d.job(id).options.networkIds.length,12);
  assert.throws(()=>d.scan('grove'),/busy/);
  assert.equal(d.snapshot('mbus').active,null);
  d.advance(3000);assert.ok(d.job(id).result.packets>0);
  assert.equal(d.job(id).result.simulated,true);
  assert.equal(d.attackStop(id),true);
  const final=d.job(id);d.advance(10000);assert.deepEqual(d.job(id),final);
  assert.equal(d.snapshot('grove').active,null);
  assert.deepEqual(d.files('grove'),[]);
 }
});

test('global Handshaker emits deterministic bounded scenario results, including an empty scenario',()=>{
 const small=structuredClone(seed);
 small.networks=small.networks.slice(0,1);
 small.clients=small.clients.filter(c=>c.network_id==='net-1');
 const a=new SimulatedDevice(small),b=new SimulatedDevice(small);
 const x=a.attackStart('grove','global_handshaker'),y=b.attackStart('grove','global_handshaker');
 a.advance(5000);for(let i=0;i<50;i++)b.advance(100);
 assert.deepEqual(a.job(x).result,b.job(y).result);
 assert.deepEqual(a.job(x).result.handshakes,[{networkId:'net-1',ssid:'NEON-BAZAAR',bssid:'02:20:77:00:00:01'}]);
 a.advance(100000);assert.equal(a.job(x).result.handshakes.length,1);
 small.networks=[];small.clients=[];
 const empty=new SimulatedDevice(small),id=empty.attackStart('grove','global_handshaker');
 empty.advance(10000);assert.deepEqual(empty.job(id).result.handshakes,[]);
});

test('global jobs cancel on disconnect/reset and do not stop another module',()=>{
 for(const kind of kinds){
  const d=new SimulatedDevice(seed),a=d.attackStart('grove',kind),b=d.attackStart('mbus',kind);
  d.setModule('grove',{connected:false});d.advance(2000);
  assert.equal(d.job(a).state,'cancelled');assert.equal(d.job(b).state,'running');
  assert.throws(()=>d.attackStart('grove',kind),/disconnected/);
  d.cancel(b);assert.equal(d.job(b).state,'cancelled');
  d.reset();assert.equal(d.job(a),null);assert.equal(d.job(b),null);
  assert.ok(d.attackStart('grove',kind)>b);
 }
});
